### DISCUSSION TITLE
How would this work with external simulation codes

### DISCUSSION THREAD

**Discussion by stefanbringuier** [rodluger/showyourwork#338] [Q&A] [answered]
I'm interested in writing a manuscript using showyourwork but its not clear how I would proceed if using simulation codes such as [LAMMPS](https://lammps.org) or [VASP](https://www.vasp.at/). These codes, which are not python based but do have wrappers, take-in text input files containing specific variable commands and values and then run computationally demanding simulations.  I have two questions:

1. What gets put into the `data` folders vs `scripts`. I get that if already have the raw simulation data that is going to be processed/analyzed for generating figures, it goes `data`, but what about the input scripts for the simulation codes that generated the raw data? Or what about the specific version/build and compile options of the simulation code, where should these go and how does this info/dependency get captured?
2. Would it be possible to specify a rule to actually generate  (optionally) the raw data if desired? This would mean providing the system path for the binary is and where in the showyourwork project the input scripts are along with any command flags.

Could be  that I'm not correctly understanding how to setup things up, as the [zenodo integration section on dynamic datasets](https://show-your.work/en/latest/zenodo/?highlight=static#dynamic-datasets) seems to point in the right direction. Just want to understand the setup of showyourwork a bit better before I invest a lot of time.

**Comment by dfm** [comment, marked as answer]
@stefanbringuier — Since `showyourwork` is built on top of `snakemake`, you can write arbitrarily complicated rules for generating any datasets or figures by specifying the rules in your `Snakefile`. The `showyourwork` docs give a brief intro to how you would do this [here](https://show-your.work/en/latest/snakefile/), but in your case you may want to look a little more closely at [the `snakemake` docs](https://snakemake.readthedocs.io/en/stable/) to see what all is possible. For example, you can write a rule like:

```python
rule simulation:
    output:
        "dataset.dat"
    shell:
        "arbitrary shell command"
```

which can run your simulation, as long as you have a way of calling it from the command line. `snakemake` also has [built in support for cluster execution using SLURM](https://snakemake.readthedocs.io/en/stable/executing/cluster.html), which should be possible to integrate with `showyourwork`.

So, in summary, the idea with the design of `showyourwork` is to expose sensible defaults for simple scripting requirements, while allowing advanced usage with the full power of `snakemake`. I hope this helps to set the context, and I'm happy to chat more!

**Reply by stefanbringuier** [reply]
@dfm  thank you very much for the reply. I see, so much of the capabilities I'm referencing falls within the functionality of `snakemake` rather than `showyourwork`.  I'll go through the docs as you recommended since I'm not familiar with `snakemake`, however I'm still unclear about the following:

1. Would the option to enable or disable running the rule be controlled by `snakemake` or `showyourwork`.
2. What is best practice for where the input scripts live for the `snakemake` shell commands, I assume it's in `src/scripts`.

Thanks!

**Reply by dfm** [reply]
Great! Responses to your questions:

> Would the option to enable or disable running the rule be controlled by `snakemake` or `showyourwork`.

`showyourwork` is designed to always run the user specified rule from the Snakefile if it is provided, otherwise it tries to fall back on the `showyourwork` default implementation. As far as I know, there isn't a built in feature for enabling or disabling these rules.

> What is best practice for where the input scripts live for the `snakemake` shell commands, I assume it's in `src/scripts`.

Exactly! I'm hoping to loosen these best practices in the future, but for now your best bet is to follow the `showyourwork` directory structure as best as you can.
